### PR TITLE
(2389) Devise two-factor login

### PR DESCRIPTION
**Pull request by rgarner**
## Changes in this PR

Two-phase Devise two-factor login. A user logs in with user name and password, and if they are set up for 2FA, receive a text message via Notify.

## Still to do:

- [x] Provision new env vars in Terraform
- [x] Adjust interval and drift for SMS ROTP
- [x] [Filter](https://github.com/tinfoil/devise-two-factor#filtering-sensitive-parameters-from-the-logs) OTP attempts from the logs

### After
Screenshot 2022-02-21 at 18 43 37

## Next steps

- [x] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [x] Do any environment variables need amending or adding?
- [x] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

**Comment by CristinaRO** [comment]
[Verify login with MFA](https://trello.com/c/4hHxXuul/2389-verify-login-with-mfa)

File: spec/lib/notify/otp_message_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require "notify/otp_message"
+
+RSpec.describe Notify::OTPMessage do
+  subject(:message) { Notify::OTPMessage.new("+4477881234567", "123456") }
```
Comment (dragon-dxw) [inline spec/lib/notify/otp_message_spec.rb:4]: If you're worried about accidently spamming someone with text messages, 07700 900000 might be a plan.

https://www.numbersupermarket.co.uk/fake-number-ranges-here-are-the-numbers-reserved-exclusively-for-tv-radio-shows/

Comment (rgarner) [reply spec/lib/notify/otp_message_spec.rb:4]: Heh! Yes, I believe the GOV.UK docs mention this explicitly!

In this case though, everything is going to the [`fake_client`](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/1589/commits/af6ca35e7c5adb30d85054367c75c296a6b8616b#diff-f39dd1bca5a438cf92e99327023c748d23a600c0545c96b26107e96f5e00a6dcR8)

Comment (rgarner) [reply spec/lib/notify/otp_message_spec.rb:4]: Yep, here they are in a page on [smoke testing](https://docs.notifications.service.gov.uk/ruby.html#smoke-testing). It's entertaining to note the provenance, mind 😆

Comment (dragon-dxw) [reply spec/lib/notify/otp_message_spec.rb:4]: Oh, didn't realise they'd got their own hooks on some of those numbers! Nice.

File: app/models/user.rb [hunk 2]
```diff
@@ -22,6 +24,10 @@ def active_for_authentication?
 
   private
 
+  def ensure_otp_secret!
+    update!(otp_secret: User.generate_otp_secret) if otp_secret.nil?
```
Comment (rgarner) [inline app/models/user.rb:28]: This is happening `before_save` so we don't need to `update!` as we'll make a second DB call; just `self.otp_secret=`

**Review by CristinaRO — APPROVED** [review]
Given that we have paired on this story, and that the whole feature will be reviewed before merging, we are confident to merge it ourselves without a third party review.

**Pull request merged**